storage_v2/v1adapter/ptrace_model.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1adapter
+
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/pkg/iter"
+	otel2model "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
+)
+
+// PTracesSeq2ToModel consumes an otel trace iterator and returns a jaeger model trace.
+//
+// When necessary, it groups chunks of traces into one single trace
+func PTracesSeq2ToModel(seqTrace iter.Seq2[[]ptrace.Traces, error]) ([]*model.Trace, error) {
+	var (
+		jaegerTraces []*model.Trace
+		err error
+		tracesByID map[pcommon.TraceID]*model.Trace
+	)
+
+	seqTrace(func(otelTraces []ptrace.Traces, e error) bool {
+		if e != nil {
+			err = e
+			return false
+		}
+
+		for _, otelTrace := range otelTraces {
+			spans := modelSpansFromOtelTrace(otelTrace)
+			for _, span := range spans {
+				traceId := span.TraceID.ToOTELTraceID()
+				if _, exists := tracesByID[traceId]; !exists {
```

Comment:
how you went into opposite direction - not _trusting_ the contact of the iterator. The contract says that if a trace is split into more than one chunk then those chunks will be _consecutive_. That means you can re-assemble the model traces without maintaining a map, simply by keeping track of the last seen trace ID.

https://github.com/jaegertracing/jaeger/blob/42a20f02b41b9b243317b4c8c4d73b2b0b087e5c/storage_v2/tracestore/reader.go#L22-L24

```
// Chunking requirements:
// - A single ptrace.Traces chunk MUST NOT contain spans from multiple traces.
// - Large traces MAY be split across multiple, *consecutive* ptrace.Traces chunks.
```

Reply:
You can use `AggregateTraces` function from `internal/jptrace/aggregator.go` together with `iter.CollectWithErrors`:
```
traces := iter.iter.CollectWithErrors(jptrace.AggregateTraces(seqTrace))
```
The `traces` is now a plain slice where each element is a complete trace that can be passed to `modelSpansFromOtelTrace`